src/server/replica.cc

```diff
@@ -781,39 +792,32 @@ void Replica::StableSyncDflyReadFb(Context* cntx) {
   io::PrefixSource ps{prefix, &ss};
 
   JournalReader reader{&ps, 0};
+  TransactionReader tx_reader{};
 
   while (!cntx->IsCancelled()) {
-    TranactionData tx_data;
-    while (!cntx->IsCancelled()) {
-      waker_.await([&]() {
-        return ((trans_data_queue_.size() < kYieldAfterItemsInQueue) || cntx->IsCancelled());
-      });
-      if (cntx->IsCancelled()) {
-        return;
-      }
-      auto res = reader.ReadEntry();
-      if (!res) {
-        cntx->ReportError(res.error(), "Journal format error");
-        return;
-      }
-      last_io_time_ = sock_->proactor()->GetMonotonicTimeNs();
-      bool tx_data_full = tx_data.UpdateFromParsedEntry(std::move(*res));
-      if (tx_data_full == true) {
-        break;
-      }
-    }
+    waker_.await([&]() {
+      return ((trans_data_queue_.size() < kYieldAfterItemsInQueue) || cntx->IsCancelled());
+    });
+    if (cntx->IsCancelled())
+      break;
+
+    auto tx_data = tx_reader.Next(&reader, cntx);
+    if (!tx_data)
+      break;
+
+    last_io_time_ = sock_->proactor()->GetMonotonicTimeNs();
+
```

Comment:
If I understood it correctly, there is actually no reason to check the queue has free space on every iteration of collecting a MULTI command (because it does not grow until we finally insert a txdata), so we can do it once per tx_data